Enhance: import/export map as JSON

SlySven opened the pull request:
This is so that a crowd sourced map might be edited in a collaborative manner. The first stage is to make sufficient of the entire map details be exported/imported in a modular (by area) fashion.

First (easier) part is writing the data out.

Include a Lua function `exportJsonMap()` to perform the export function.

Also:
* Rename `(QMultiMap<int, QPair<int, int>>) TArea::exits` to: `TArea::mAreaExits` because the old name was such a common word in our source code it was hard to find the uses of this member.

This format has been constructed so as to not mention the most common, or default values for some items - so as to minimise details that have to be included on the basis they can be assumed when reconstructing them on the other end.

On the other hand the whole file is compressible so for storage (but not for diff/git work) archiving / compressing the file is recommended! For instance a binary map file I have is 18.6MB which produced a 25.8MB JSON file which I was able to compress down to 2.9MB - obviously this is *very* content dependent, so other's Miles-May-Vary...

As the export process is not that fast include a progress dialogue that shows how many areas, map labels and rooms have been processed into the JSON format. For a 20K room map with 40 odd areas and around 800 map labels
(which are awkward to convert to a text-like form) this can take 30 seconds on my 1.8GHz 4 Core PC!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Enhance: import/export map as JSON

This is so that a crowd sourced map might be edited in a collaborative
manner. The first stage is to make sufficient of the entire map details
be exported/imported in a modular (by area) fashion.

First (easier) part is writing the data out.

Include a Lua function `exportJsonMap()` to perform the export function.

Also:
* Rename `(QMultiMap<int, QPair<int, int>>) TArea::exits` to:
`TArea::mAreaExits` because the old name was such a common word in our
source code it was hard to find the uses of this member.

This format has been constructed so as to not mention the most common, or
default values for some items - so as to minimise details that have to be
included on the basis they can be assumed when reconstructing them on the
other end.

The other hand the whole file is compressible so for storage (but not for
diff/git work) archiving / compressing the file is recommended! For
instance a binary map file I have is 18.6MB which produced a 25.8MB JSON
file which I was able to compress down to 2.9MB - obviously this is *very*
content dependent, so other's Miles-May-Vary...

As the export process is not that fast include a progress dialogue that
shows how many areas, map labels and rooms have been processed into the
JSON format. For a 20K room map with 40 odd areas and around 800 map labels
(which are awkward to convert to a text-like form) this can take 30
seconds on my 1.8GHz 4 Core PC!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: keep CodeFactor happy

It spotted a missing space between an `ìf` and the opening brace of it's
conditions.

It had a recommendation about a constant that I was using to set the
dimension of a `char[]` (array) - it felt a compile time `constexpr` was
a better thing to use.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:construction_worker: :hammer_and_wrench: Note this is currently a Work In Progress and there is no guarantee that the data it produces is enough to reproduce your map. Hell there isn't even a reader for the file produced yet - that is the next step...

Also - the exact format might change once I try to construct that reader (if/when I find I have done this rong.)

SlySven committed:
BugFix: make code build-able for pre Qt 5.14 versions

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFixBugFix: fix mistake in previous fix!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: more pre Qt 5.14 version fixes

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Update: make Cancel button work in big areas

Although the existing code would abort at the end of an area, for some
humongous maps with a few very large areas it is also a good idea to check
for the cancel button being pressed each time the progress bar is updated.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: correct leaving an extra } at the end of a source file.

I was trying to avoid committing some code that I was already working on
but not ready to push out and accidentally left a fragment (a closing `}`)
in place which broke compilation...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
@wilk you have experience working with the map data format, can you help in reviewing?

wilk commented:
Yeah, @vadi2 , I'd like to but I'm sure you're looking for someone else 😄

Delwing commented:
@vadi2 sure, but I guess it's not done yet :) I've asked SlySven about estmates, since I have deep interest in that feature :)
Will keep my on that :)

SlySven committed:
W.I.P. start on reader

Renamed `(QMap<int, int>) TMap::envColors` to `TMap::mEnvColors`
and `(QMap<int, QColor>) Tmap::customEnvColors` to `TMap::mCustomEnvColors`
so that it is clearer that they are members of the `TMap` class.

Add alpha component to end of list of (now four) 0 to 255 integer values
returned by `getCustomEnvColorTable()` - as the corresponding setter does
allow one to be provided.

Extract some of the `QStringLiterals` in `./src/TMap.cpp` to predefined
constants so that only one of each string is created/needed in that file.

Fixed handling of `TMap::mCustomEnvColors` as it was not being written
correctly. Version incremented to `0.002` - and only that form will be
read currently. The reader is incomplete at present - only a test read of
the direct TMap elements are loaded to temporary variables for testing.
The area (and their contained rooms) are still to have parser created.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:warning: The code so far is flawed in that it is not populating the **"int (color Id)"** => **R,G,B(,A) trip/quadrup-let color** **`"customEnvColors"`** table.

The **`"int (IRE MUD color code)"`** => **`ìnt (Mudlet color Id)`** **`"envColor"`** table is probably also not being filled in either...

SlySven committed:
Merge branch 'development' into Enhance_importExportMapAsJSON

Needed to bring in changes to the way special exits and special exit locks
are stored in the `TRoom` class.

Conflicts resolved in:
* src/TArea.cpp
* src/TMap.cpp
* src/TRoom.cpp
* src/TRoom.h

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Complete: provide JSON importer

Move initialisers for TMap, TRoomDB, TRoom and TArea to header file where
possible. As per Issue #4578.

Move the default and unnamed area names from TRoomDB to TMap - as it made
setting them up easier (though one of them does need to be initilised
before the normal TRoomDB instance associated with the TMap is itself
initialised. This meant putting these private members near the top of the
header file even though we normally put private ones down the bottom.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Actually it is waiting for #4604 to be merged into the `development` branch.

SlySven committed:
Prepare: revise so that development branch can be merged more cleanly

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_importExportMapAsJSON

Conflicts resolved in:
* src/T2DMap.cpp
* src/TArea.cpp
* src/TArea.h
* src/TLuaInterpreter.cpp
* src/TMap.cpp
* src/TMapLabel.h

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: add handling of Room Symbols colour

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Refactor: move JSON colour writing/reading code from TArea/TRoom to TMap

The code is common to all three classes so can be shared. At the same time
make it explicit in the key as to whether there is an alpha component so
the colour is a 24Bit opaque one or a 32Bit one with transparency.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: initialise mDlgRoomSymbol so that settng room symbols works again (#4687)

A test for this pointer being a `nullptr` is what enables a new dialogue to
be made in the 2D mapper - by not initialising it in #4573 a random
non-null value is found so the dialogue and thus the functionality can be
virtually impossible to use.

As it is a pointer I have taken the liberty to introduce a `p` into the
initial part of the name - after the `m` to indicate it is a class member.

As the defect is a regression that causes loss of functionality I am rating
this as a **high** priority bug(fix) even though it does not induce any
fatal termination of the application.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fixup: ensure code used in this PR matches the development branch version

In creating #4693 I chose the opposite duplicate chunk of code that I was
using in this PR - since there was slight differences it would have
impacted when this PR is merged.

This commit swaps things so that it uses the same code as will be in
the development branch in the near future.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: add distinctive prefix to all the new static const values

During a CI build of a previous commit it failed on one platform because
one of the `static const` I had used `SIZE` was clashing with something
else. Also I was noticing that `ID` was lighting up in my Linux `qtcreator`
as if it was a member of one of the map classes - yet it was not one that
I could see defined anywhere. To avoid mixups I have added a `sc` for
`static const` prefix - which does fit with some of the other prefixes we
are using.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Enhance_importExportMapAsJSON

Conflicts resolved in:
* src/T2DMap.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: various fixups

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
Right this seems to be functional, use:
* `exportJsonMap("path/to/fileName.json")` to write out a complete copy of all the map data to a filename (which will be given a `.json` if one is not on the end of the file name
* `importJsonMap("path/to/fileName.json")` to read in such a file.

Both will return true on success or `nil` + an error message on failure. There is some validation and an embedded major.minor version scheme - and the current reader expects a version of `0.003` only.

I was considering adding a second argument to the writer - something along the lines of `enforceAlphabeticalElementOrdering` - this is because the JSON standard is, like Lua unordered in the way it stored keys. However the current Qt implementation does order them; hopefully they do not drop this for Qt 6, but if they do we will have to change the way the output is generated (to using purely *arrays* which have a strict ordering of their contents compared to *objects* which do currently but need not). This putative argument would switch to the array format but I haven't fully worked out the details and do not look forward to having to do so - as it is such a change would not be helpful to anyone using this to track changes in a map in a visible manner like what I am expected it to be used for.

vadi2 committed:
Merge remote-tracking branch 'origin/development' into Enhance_importExportMapAsJSON

vadi2 commented on src/TMap.cpp:
```diff
@@ -2649,6 +2653,542 @@ void TMap::setRoomNamesShown(bool shown)
+    mpProgressDialog->setMinimumWidth(500);
+    mpProgressDialog->setAutoClose(false);
+    mpProgressDialog->setAutoReset(false);
+    mpProgressDialog->setMinimumDuration(0); // Normally waits for 4 seconds before showing
```
In case you have a tiny map, the dialog will flash in front of the eyes for no reason - while 1s is a good amount of time to wait without seeing progress:

```suggestion
    mpProgressDialog->setMinimumDuration(1);
```

SlySven replied:
I'll check this out and see...

SlySven replied:
... I've put it into the next commit.

vadi2 commented on src/TMap.cpp:
```diff
@@ -2649,6 +2653,542 @@ void TMap::setRoomNamesShown(bool shown)
+    QFile file(destination);
+    if (!file.open(QFile::OpenMode(QFile::Text|QFile::WriteOnly))) {
+        qWarning().noquote().nospace() << "TMap::writeJsonMapFile(...) WARNING - Could not open save file \"" << destination << "\".";
+        mpProgressDialog->setAttribute(Qt::WA_DeleteOnClose, true);
```
Shouldn't this be set a creation?

SlySven replied:
:thinking: TBH I am not sure - I can't recall whether the dialogue is decorated with a title bar with a close button, need to see if it has and then how things behave if I close it prematurely...

SlySven replied:
It *seems* that the close button on the titlebar is the same as the "Cancel" button...

vadi2 commented on src/TMap.cpp:
```diff
@@ -2649,6 +2653,542 @@ void TMap::setRoomNamesShown(bool shown)
+    // Use this to track any changes. in a major.minor number format, minor
+    // is to be three digits long.
+    // 0.002 was the first published draft
+    // 0.003 changed the format to encapsulte the room symbol text, it also
```
This is a tough sentence to read

SlySven replied:
Trying to keep track of what I changed - and might have broken in each iteration. I'll see if I can be more readable.

SlySven replied:
I've tarted this up a bit and moved it to the start of the method.

vadi2 commented on src/TRoom.cpp:
```diff
@@ -59,36 +63,53 @@ QDataStream &operator>>(QDataStream& ds, Qt::PenStyle& value)
     return ds;
 }
 
+static const QColor scDefaultHighlightForeground(QColor(255, 150, 0));
```
Some of these are only used twice. I don't think the benefit is worth the difficulty of reading the code they add.

SlySven replied:
At the time I started I wasn't sure of how many times I would need each string, or in this case, colour.

vadi2 replied:
I understand, but now would be good to clean them up. They do make code difficult to read and there's actually no benefits to using them like that

SlySven replied:
Will see what I can do...

vadi2 commented on src/TArea.cpp:
```diff
@@ -584,3 +606,372 @@ int TArea::createLabelId() const
+          && pLabel->bgColor.red() == defaultLabelBackground.blue())) {
+
+        // For an image the colors are not used and tend to be set to black, if
+        // so skip them. Ufortunately because of the way QColour s are assembled
```
```suggestion
        // so skip them. Unfortunately because of the way QColours are assembled
```

Same with other cases

SlySven replied:
:blush:

SlySven replied:
Fixed in next commit.

vadi2 reviewed (commented):
Amazing amount of work! 

Colors are lost on export/import: https://streamable.com/r9u16j

SlySven commented:
> Colors are lost on export/import: https://streamable.com/r9u16j

I can't view that content.

Does it look like an export or an import problem - are there `color_24RGB` or `color_32RGBA` elements in the file for the things that are missing colours or not?

vadi2 commented:
Haven't dug into what goes wrong, the map is available here: https://github.com/IRE-Mudlet-Mapping/AchaeaCrowdmap/tree/gh-pages/Map area is `Mhaldor Isle`. Map before export:

![image](https://user-images.githubusercontent.com/110988/106354157-e6928580-62ef-11eb-88f6-54db02e07ee4.png)

Map after import:

![image](https://user-images.githubusercontent.com/110988/106354165-ee522a00-62ef-11eb-8cc0-fff6623c43ba.png)

It's not always all red, sometimes the colors change up randomly.

SlySven commented:
Humm, the IRE environment\<==>colour remapping is being used for your map @vadi2 . I can load that file (BTW why doesn't it have a `.dat` extension?) - and go to the room you selected - **10577** but that remapping confuses the hell out of colour selection - note how that grey-green is shown as colour **10** in the room colour selector - but that is a shade of cyan:
![image](https://user-images.githubusercontent.com/6163092/106360217-2ae14e00-630f-11eb-91b3-88f890419da0.png)
```lua
lua getRoomEnv(10577)
12
```
but something is indeed different if the map is cleared (with a Lua `loadMap("_")`) and then an exported Json map is reimported:
![image](https://user-images.githubusercontent.com/6163092/106360323-bd81ed00-630f-11eb-95ff-d06f1c4bda5b.png)
```lua
lua getRoomEnv(10577)
12
```

BTW Some of the areas in the original crowdmap file have area names that are empty strings - but I thought we prohibited this nowadays. Certainly the audit after the `importJsonMap(...)` call fixes this:
![image](https://user-images.githubusercontent.com/6163092/106360503-a1cb1680-6310-11eb-9efc-971e91ade220.png)

vadi2 commented:
> BTW Some of the areas in the original crowdmap file have area names that are empty strings - but I thought we prohibited this nowadays. 

It's exported from a modern Mudlet, I suspect there are bugs in the export still.

SlySven commented:
Also - those areas (like the one with id: **`88`**) are missed from the output of both `getAreaTable(...)` {which has the area name as key - which one good reason to prohibit duplicate area names} and `getAreaTableSwap(...)` {which has the area id as key}!

SlySven commented:
I do not understand why it was not picked up when the Mudlet binary map (format **20**) crowdmap was read though...

SlySven commented:
Ah, found the reason `QMap<int, int>TMap::mEnvColors` (was called `TMap::envColors`) was not being repopulated during `importJsonMap(...)` call - need to find out why.

SlySven commented:
... oh :hankey: - forgot to put a *working* reader in for it - there is one but it is borked!

SlySven committed:
BugFix: correct code that was failing to parse environment-colour mapping

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
Looks better now! 

@Delwing want to have a look, as it'll be useful for the web based map viewer?

Delwing commented:
Sure, definitely will replace custom export with that one :) Will test and review some time later.

SlySven committed:
Revise: peer-review items and other tweaks

Note that this revises the format version to be 1.000 (ready for release)
so, although the format has not changed, any recent files produced during
evaluation will need to be hand edited to change the line:
    "formatVersion": 0.003,
to:
    "formatVersion": 1.000,
in order to read them now.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

ryanstadther commented:
@vadi2 and @SlySven 

Not sure why this is happening but when I try "goto port" with original crowdmap I get this
![image](https://user-images.githubusercontent.com/47157604/106367567-29148c00-6309-11eb-8553-e7d89ffc382b.png)

But when I try to export json and re import I get a different order?
![image](https://user-images.githubusercontent.com/47157604/106367582-45b0c400-6309-11eb-9992-d14a594ce78c.png)

ryanstadther commented:
Also I'm having issues with special exits, here is a spot in Petra where you need to "push gate" to go west but I just get a strange question mark...

![image](https://user-images.githubusercontent.com/47157604/106367685-ff0f9980-6309-11eb-81a7-de96388456d7.png)

![image](https://user-images.githubusercontent.com/47157604/106367703-18b0e100-630a-11eb-8d43-9509efa0646e.png)

ryanstadther commented:
got this warning too while trying the special exits
![image](https://user-images.githubusercontent.com/47157604/106367735-51e95100-630a-11eb-815c-be101bff23d9.png)

ryanstadther commented:
Looks like when the room color is black then the arrows to another area are also black and you can't see them, in the current room I am standing there is a black arrow to a new area going south, not sure if there could be a check in there if room color = background color to change it to white maybe?

![image](https://user-images.githubusercontent.com/47157604/106368259-f9b44e00-630d-11eb-9203-db1a751a6e0b.png)

ryanstadther commented:
in setting mapper colors I only see grey?
![image](https://user-images.githubusercontent.com/47157604/106368299-341deb00-630e-11eb-812f-ac4ef8f1b5d2.png)

Also I tried changing the background color to white to see how it looked on the map, but it changed the main display to ..partially white?...and now it wont change back :(
![image](https://user-images.githubusercontent.com/47157604/106368319-5283e680-630e-11eb-89f8-a9b3370ad7f7.png)

SlySven commented:
TBH I cannot make head nor tail of most of your errors. This PR should enable the same map data as was present on exporting to the replicated when it is imported.

In https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770278308 the room that is *selected* (not the one with the red mark around it but with the blue diagonal gradient) **7615** has north (to 7616), north-west (to 7618), west (to 7617) south-west (to 37178) and east (to 7613) exits. The room that the player is in (with the red mark) is **7613** and that has a north exit (to 7614) and a south exit (7112) and two special exits ("push gate" to 7615 and "worm warp" to 6956).

I have a completely different speedwalking function in my profile but I agree that the data it gives for a walk from 7613 to 7615 is odd - indeed it is the inverted question mark. However I can get that using the original map.

Ah! You have spotted a real-live :beetle: - and I have worked out what caused it - it was introduced by #4526 .

*Edit: fix pending in #4722 ...*

ryanstadther commented:
@SlySven Ahh yeah, could be unrelated to the json stuff I guess...I don't know how things work! :)

SlySven commented:
Is your OS Windows? I thought I spotted something like https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770282292 in my MSYS2/Mingw-w64 local builds but I didn't know that it was wider than just my setup.

*Edit - see Issue #4743*

ryanstadther commented:
> Is your OS Windows? I thought I spotted something like [#4546 (comment)](https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770282292) in my MSYS2/Mingw-w64 local builds but I didn't know that it was wider than just my setup.

Yep Windows

SlySven commented:
https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770282025 is going to happen whenever the room colour matches the background colour - not sure that there needs to be a work around if it can be solved by sensible configuration...

ryanstadther commented:
> [#4546 (comment)](https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770282025) is going to happen whenever the room colour matches the background colour - not sure that there needs to be a work around if it can be solved by sensible configuration...

Yeah I agree...I have my "black" as actually a dark grey so that works.

SlySven committed:
Merge branch 'development' into Enhance_importExportMapAsJSON

Needed to pull in fix for special exits in speed-walks.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
> https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770282025

Agree with the idea of the check, it should "just work"

SlySven commented:
> Agree with the idea of the check, it should "just work"

Yeah, but it is out of scope of *this* PR I think.

*Edit: added as Issue #4764.*

vadi2 commented:
@ryanstadther just curious do you get what the https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770278483 warning is trying to tell you?

ryanstadther commented:
@vadi2 no clue

SlySven committed:
Merge branch 'development' into Enhance_importExportMapAsJSON

Needed to include fix for crashing on exit in TTimerUnit destructor.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
> @vadi2 no clue

*\*sigh\**

The text that was to be sent to the MUD to move in that special exit direction turns out to **really** be `¿` U+BF {INVERTED QUESTION MARK} but that is not something that ASCII could encode and send to the MUD Server.

Imaging you had a system to tell (a non-blind person - this isn't an analogue about "explaining colour to a blind {from birth} person!") someone about colour in the **visible** light range - what would happen if you tried to tell them about something that is ultra-violet or infra-red?

It is impossible - and that is the sort of situation that message is (failing obviously) to convey!

SlySven committed:
Merge branch 'development' into Enhance_importExportMapAsJSON

vadi2 commented on src/TMap.h:
```diff
@@ -128,32 +137,50 @@ class TMap : public QObject
+    std::pair<bool, QString> writeJsonMapFile(const QString&);
+    std::pair<bool, QString> readJsonMapFile(const QString&);
+    int getCurrentProgressRoomCount() const { return mProgressDialogRoomsCount; }
+    bool incrementJsonProgressDialog(const bool isExportNotImport, const bool isRoomNotLabel, const int increment = 1);
```
![image](https://user-images.githubusercontent.com/110988/107112775-74bfbc00-685a-11eb-9761-251933aae498.png)

This hurts my brain... how about just a simple struct instead?

SlySven replied:
It evolved like that - each argument got added at a separate stage. As it is ***you*** are using some fancy IDE that is documenting the argument names anyhow for these PODs so why complicate things with a `struct`ure which will have to be initialised before being passed anyhow...

vadi2 replied:
Sorry, I mean a flag from a struct. This evolved code is hard to read, could be improved.

Otherwise just waiting on the issues reported to be resolved, need any help with them?

SlySven replied:
> Sorry, I mean a flag from a struct. This evolved code is hard to read, could be improved.

I don't get what you mean - could you suggest what you think you want?

:confused: Um, what issues are left that are germane to this PR:
* https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770277594 is, at best guess, down to a table being stored in lua by code outside of my arena and the order there is random is it not?
* https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770278308 has been diagnosed and fixed already in #4722 and was the cause of:
* https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770278483 that is not a feature of this PR and it is being handled elsewhere in #4741.
* https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770282025 is also not a part of this PR and I have raised an issue #4764 to track it.
* https://github.com/Mudlet/Mudlet/pull/4546#issuecomment-770282292 the first part is also not part of this PR was already known about as #4743 is being resolved by yourself in #4756 - the second part is normal operation as far as I can tell - changing the main background colour does (I think) reset the default background colour for any newly generated text but does not get applied retrospectively to the existing text.

vadi2 commented on src/TArea.cpp:
```diff
@@ -40,7 +42,26 @@
+static const QColor defaultLabelBackground(QColor(0, 0, 0));
+static const int kPixmapDataLineSize = 64;
+
+static const QString scCOLORS{QStringLiteral("colors")};
+static const QString scCOORDINATES{QLatin1String("coordinates")};
+static const QString scGRID_MODE{QStringLiteral("gridMode")};
+static const QString scID{QStringLiteral("id")};
+static const QString scIMAGE{QStringLiteral("image")};
+static const QString scLABELS{QStringLiteral("labels")};
+static const QString scNAME{QStringLiteral("name")};
+static const QString scROOM_COUNT{QStringLiteral("roomCount")};
+static const QString scROOMS{QStringLiteral("rooms")};
+static const QString scSCALED{QStringLiteral("scaled")};
+static const QString scSHOW_ON_TOP{QStringLiteral("showOnTop")};
+static const QString scSIZE{QLatin1String("size")};
+static const QString scTEXT{QStringLiteral("text")};
+static const QString scUSER_DATA{QStringLiteral("userData")};
```
These are a pain and make the code difficult to read. Also, since we're creating a QString, there are no advantages to `QStringLiteral` it turns out. I recommend removing these - I understand it's good to have for consistency, but honestly in our small codebase the error is unlikely to happen.

SlySven replied:
Well, we are creating one `QString` (per compilation unit) instead of one for each use within that CU (.cpp file). i will have a look to see whether `QLatin1String` offers a compromise (again there will likely be only one copy per CU as the compiler should share them) so the run-time construction cost should not be that much even with multiple uses through the file.

SlySven replied:
Humm, `QJsonObject` and related has methods that have `QLatin1String` overloads - but they were only introduced in Qt 5.14 - so if I read that right they are reasonable to use given that all our CI builds are using Qt 5.15 and the extra cost of the more complicated construction prior to that is just going to be a burden on systems not at the leading edge of Qt.

vadi2 replied:
Sorry, but the developer pain that this code is adding is not worth the convenience it promises. I think if you run the numbers, you won't even see a 5% improvement.

vadi2 commented on src/TArea.cpp:
```diff
@@ -584,3 +606,372 @@ int TArea::createLabelId() const
     return labelId;
 }
+
+void TArea::writeJsonArea(QJsonArray& array) const
```
Constants excluded, nice and easy to follow :+1:

vadi2 commented on src/TArea.cpp:
```diff
@@ -584,3 +606,372 @@ int TArea::createLabelId() const
+    }
+
+    QStringList keys = obj.keys();
+    for (int i = 0, total = keys.count(); i < total; ++i) {
```
Difficult to follow - can a simple ranged-for be used to read the object's key and value at once? This pattern is repeated elsewhere and should be simplified.

vadi2 commented on src/TArea.cpp:
```diff
@@ -584,3 +606,372 @@ int TArea::createLabelId() const
+    const QJsonValue imageValue{imageArray};
+    labelObj.insert(scIMAGE, imageValue);
+
+    // (bool) pLabal->highlight is not saved as it is only used during editing
```
```suggestion
    // (bool) pLabel->highlight is not saved as it is only used during editing
```

vadi2 commented on src/TLuaInterpreter.cpp:
```diff
@@ -16965,3 +16977,63 @@ int TLuaInterpreter::showNotification(lua_State* L)
+    }
+
+    if (!lua_isstring(L, 1)) {
+        lua_pushfstring(L, "exportJsonMap: bad argument #1 type (export pathFileName as string expected, got %s!)", luaL_typename(L, 1));
```
Could you update to the new style functions, below as well?

Kebap replied:
Added suggestions for your convenience

vadi2 commented on src/TMap.cpp:
```diff
@@ -2649,6 +2653,550 @@ void TMap::setRoomNamesShown(bool shown)
+        return {false, QStringLiteral("no areas detected")};
+    }
+
+    mProgressDialogAreasTotal = qRound(mapObj[scAREA_COUNT].toDouble());
```
Why all of this conversion? This is something that could use a comment

SlySven replied:
All numbers (even integers) are `doubles` in JSON aren't and this seemed to me to be the most robust means of reading something that *could* have been edited by a user (or even a third party tool).

vadi2 commented on src/TMap.cpp:
```diff
@@ -2649,6 +2653,550 @@ void TMap::setRoomNamesShown(bool shown)
+    }
+
+    QMap<int, int> envColors;
+    if (mapObj.contains(scENV_TO_COLOR_MAPPING) && mapObj.value(scENV_TO_COLOR_MAPPING).isObject()) {
```
This is just reading a simple key:value from the json but the code is not simple - could a range-for be used to simplify things here? Same with custom colors below, takes some effort to parse what it is doing - when it shouldn't.

SlySven replied:
It is trying to be robust - first it checks for the thing to exist, then it checks that it is the right *sort* of thing (object vs array) - in other places where it is an array it also checks that it has a reasonable number of elements. I've seen other JSON stuff in a few places and it too is checking that something is not undefined before it tries to do anything doesn't it?

vadi2 replied:
Yeah but if we can get a ranged for, it won't even trigger if there's nothing there?

vadi2 reviewed (commented):
Works great! Nice piece of work :)

The only thing I'd improve on is that the object reading code is possibly more complicated than it should be: trying to read the keys first, then iterating them using and old-style iterator. Can this be improved with a ranged-for pattern that would be much simpler to read?

Kebap commented:
> Include a Lua function `exportJsonMap()` to perform the export function.

Do we have a button? Also for import? Maybe in preferences next to existing backup? Sorry did not test, yet.

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -16965,3 +16977,63 @@ int TLuaInterpreter::showNotification(lua_State* L)
+    }
+    QString dest = lua_tostring(L, 1);
+    if (dest.isEmpty()) {
+        lua_pushnil(L);
+        lua_pushstring(L, "a non-empty path and file name to write to must be provided");
+        return 2;
```
```suggestion
        return warnArgumentValue(L, __func__, QStringLiteral("a non-empty path and file name to write to must be provided"));
```

SlySven replied:
Fair enough - but this started life before you got to work in there...

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -16965,3 +16977,63 @@ int TLuaInterpreter::showNotification(lua_State* L)
+    }
+
+    if (auto [result, message] = pHost->mpMap->writeJsonMapFile(dest); !result) {
+        lua_pushnil(L);
+        lua_pushfstring(L, message.toUtf8().constData());
+        return 2;
```
```suggestion
        return warnArgumentValue(L, __func__, message);
```

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -16965,3 +16977,63 @@ int TLuaInterpreter::showNotification(lua_State* L)
+{
+    Host* pHost = &getHostFromLua(L);
+    if (!pHost || !pHost->mpMap || !pHost->mpMap->mpMapper || !pHost->mpMap->mpMapper->mp2dMap) {
+        lua_pushnil(L);
+        lua_pushstring(L, "importJsonMap: no map present or loaded!");
+        return 2;
```
```suggestion
        return warnArgumentValue(L, __func__, QStringLiteral("no map present or loaded"));
```

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -16965,3 +16977,63 @@ int TLuaInterpreter::showNotification(lua_State* L)
+        return 2;
+    }
+
+    if (!lua_isstring(L, 1)) {
+        lua_pushfstring(L, "importJsonMap: bad argument #1 type (import pathFileName as string expected, got %s!)", luaL_typename(L, 1));
+        return lua_error(L);
+    }
+    QString source = lua_tostring(L, 1);
```
```suggestion
    QString source = getVerifiedString(L, __func__, 1, "import pathFileName");
```

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -16965,3 +16977,63 @@ int TLuaInterpreter::showNotification(lua_State* L)
+    }
+    QString source = lua_tostring(L, 1);
+    if (source.isEmpty()) {
+        lua_pushnil(L);
+        lua_pushstring(L, "a non-empty path and file name to read to must be provided");
+        return 2;
```
```suggestion
        return warnArgumentValue(L, __func__, QStringLiteral("a non-empty path and file name to read to must be provided"));
```

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -16965,3 +16977,63 @@ int TLuaInterpreter::showNotification(lua_State* L)
+    }
+
+    if (auto [result, message] = pHost->mpMap->readJsonMapFile(source); !result) {
+        lua_pushnil(L);
+        lua_pushfstring(L, message.toUtf8().constData());
+        return 2;
```
```suggestion
        return warnArgumentValue(L, __func__, message);
```

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -16965,3 +16977,63 @@ int TLuaInterpreter::showNotification(lua_State* L)
+{
+    Host* pHost = &getHostFromLua(L);
+    if (!pHost || !pHost->mpMap || !pHost->mpMap->mpMapper || !pHost->mpMap->mpMapper->mp2dMap) {
+        lua_pushnil(L);
+        lua_pushstring(L, "exportJsonMap: no map present or loaded!");
+        return 2;
```
```suggestion
        return warnArgumentValue(L, __func__, QStringLiteral("no map present or loaded"));
```

Kebap commented on src/TLuaInterpreter.cpp:
```diff
@@ -16965,3 +16977,63 @@ int TLuaInterpreter::showNotification(lua_State* L)
+        return 2;
+    }
+
+    if (!lua_isstring(L, 1)) {
+        lua_pushfstring(L, "exportJsonMap: bad argument #1 type (export pathFileName as string expected, got %s!)", luaL_typename(L, 1));
+        return lua_error(L);
+    }
+    QString dest = lua_tostring(L, 1);
```
```suggestion
    QString dest = getVerifiedString(L, __func__, 1, "export pathFileName");
```

Kebap commented on src/TRoom.h:
```diff
@@ -131,61 +133,91 @@ class TRoom
     QMap<QString, bool> customLinesArrow;
-    bool highlight;
+
+    bool highlight= false;
```
```suggestion
    bool highlight = false;
```

SlySven replied:
Done in next commit.

Kebap reviewed (commented):
A few suggestions from me 😄

SlySven committed:
Merge branch 'development' into Enhance_importExportMapAsJSON

Conflicts resolved in:
* src/TLuaInterpreter.cpp
* src/TLuaInterpreter.h
* src/dlgMapper.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: apply new standardised error handling methods to new code in the PR

This is so that the two new methods adopt the same new code that other
functions in the TLuaInterpreter class do.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: feedback from peer review

Revert use of a single common `static const QString` for each string needed
in the JSON map file handling in `TMap`, `TArea` and `TRoom` classes.

Switch to "range based" for-loops for some of the JSON additions.

Fix a spelling mistake in a comment.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Fixup: ensure partially built new TRoomDB is destroyed if reading aborted

Not doing this would cause a resource leak if the abort button was clicked
during importation of a Json map file.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: make CodeFactor happier

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
@vadi2 How is it looking now? I've taken out those `static const QString`s and adopted range-based `for`s where I could see how to do so.

vadi2 reviewed (approved):
Thank you! It looks great on my side :1st_place_medal: 

@Kebap  had some good comments in there about making it possible to use json from the interface, will defer final approval to him.

SlySven commented:
> @Kebap had some good comments in there about making it possible to use json from the interface, will defer final approval to him.

Yeah - I understood the desire for that - but TBH this is big enough already and I wanted to get this in as something that can be used before I go on to:
* Add it to the UI - probably as an additional file type on the save/load buttons.
* Research and add Lua API to *swap* one or more areas with replacements from JSON file(s) - and the means to export them - gotta work out the logistics which would probably include a requirement for all the exits from each half (the map to received the import less the areas to be swapped out + the replacement areas to be swapped in) to connect up to rooms in the other part.

Also, as a more general issue, I think we ought to have a method to control which user data items are saved/exported so that the user can specify that a particular data item is:
* exported in all cases (*"Public"*)
* only exported to locally shared maps (*"Protected"*)
* never exported (*"Private"*)

Obviously the autosave at the end of a session - and the default - must be to save all data; but when sharing the map with other profile/users there is a requirement to be a bit more selective. As it happens this is something that I have been conscious of for a while - and your (@Vadi) observation that each characters' current room is saved in the file; and is now visible in a human readable form in the JSON export is very related to this as well. Fortunately the detail can be found and hand-edited out should anyone not like it - but the data that is leaked is only the local name that each profile has and the location that was recorded when the file was last updated for that profile - and for a map file that is shared/copied to another profile the data will be stale and out of date as soon as the profile on the machine where the player location is actively used changes (humm, does that sentence makes sense to *anyone*...? :dizzy:)

vadi2 commented:
> TBH this is big enough already and I wanted to get this in as something that can be used before I go on to

I see what you mean, I agree. @Kebap @Delwing ?

Delwing commented:
I've posted question while ago - why do we need to export room highlights if they are not permanent normally.

SlySven commented:
> I've posted question while ago - why do we need to export room highlights if they are not permanent normally.

Ah, well spotted - I'll check that, and if so, I've whip that detail out...

SlySven commented:
:thinking: Confirmed - ***none*** of the room high-lighting details (size, two colours and state) are stored in the binary map format!

Is that the wanted state of things? I'll comment out the writer method but leave the reader active.

SlySven committed:
Revise: disable writing out Room highlighting details

It has been pointed out that the binary map format does not save the room
highlighting details either - so replicate that behaviour for the moment.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

Kebap commented:
> > making it possible to use json from the interface
> 
> Yeah - I understood the desire for that - but TBH this is big enough already and I wanted to get this in as something that can be used before I go on to:
> 
> * Add it to the UI - probably as an additional file type on the save/load buttons.

Seems fair to me, approved!

SlySven commented:
:books: Wiki documentation added and links inserted into `TLuainterpreter.cpp` at (2) appropriate places #4810.
